src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/OffenderKeyDatesService.kt

```diff
@@ -3,20 +3,58 @@ import arrow.core.getOrElse
 import org.springframework.http.HttpStatus
 import org.springframework.stereotype.Service
 import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.enumerations.CalculationStatus
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.enumerations.ReleaseDateType
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.exceptions.CrdWebException
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.CalculationContext
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.NomisCalculationSummary
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.OffenderKeyDates
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.ReleaseDate
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.ReleaseDatesAndCalculationContext
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.repository.CalculationRequestRepository
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.resource.CalculationController
+import java.util.*
 
 @Service
 @Transactional(readOnly = true)
 class OffenderKeyDatesService(
   private val prisonService: PrisonService,
   private val calculationResultEnrichmentService: CalculationResultEnrichmentService,
+  private val calculationRequestRepository: CalculationRequestRepository,
 ) {
 
+  fun getKeyDatesByCalcId(calculationRequestId: Long): ReleaseDatesAndCalculationContext {
+    try {
+      val calculationRequest = calculationRequestRepository.findById(calculationRequestId).get()
+      val offenderKeyDatesEither = prisonService.getOffenderKeyDates(calculationRequest.bookingId)
+
+      return offenderKeyDatesEither
+        .map { releaseDates(it) }
+        .map { calculationResultEnrichmentService.addDetailToCalculationDates(it, null, null).values.toList() }
+        .fold(
+          { throw NoSuchElementException("Unable to retrieve offender key dates") },
+          { enrichedDates ->
+            ReleaseDatesAndCalculationContext(
+              CalculationContext(
+                calculationRequestId,
+                calculationRequest.bookingId,
+                calculationRequest.prisonerId,
+                CalculationStatus.CONFIRMED,
+                calculationRequest.calculationReference,
+                calculationRequest.reasonForCalculation,
+                calculationRequest.otherReasonForCalculation,
+                calculationRequest.calculatedAt.toLocalDate(),
+                calculationRequest.calculationType,
+              ),
+              enrichedDates,
+            )
+          },
+        )
+    } catch (e: Exception) {
+      throw CrdWebException("Unable to retrieve offender key dates", HttpStatus.NOT_FOUND)
```

Comment:
worth logging something do you think as original cause is lost? Also if the only thing that's throwing is the fold why not just throw CrdWebException straight away instead of catch and re-throw?

Reply:
The catch also covers error in other parts of the method, so i have logged the error before throwing the custom exception.